source/microbit/modaudio.cpp

```diff
@@ -0,0 +1,430 @@
+    audio_stop();
+    pin0 = p0;
+    pin1 = p1;
+    audio_gpiote_init(pin0, 0);
+    audio_gpiote_init(pin1, 1);
+    audio_ppi_init(0, 1);
+}
+
+void audio_set_pins(mp_obj_t pin0_obj, mp_obj_t pin1_obj) {
+    PinName p0 = microbit_obj_get_pin(pin0_obj)->name;
+    PinName p1 = microbit_obj_get_pin(pin1_obj)->name;
+    if (p0 != pin0 || p1 != pin1) {
+        init_pins(p0, p1);
+    }
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(microbit_audio_set_pins_obj, audio_set_pins);
+
+mp_obj_t audio_reset(void) {
+    init_pins(pin0, pin1);
+    memset(audio_buffer_ptr, 0, AUDIO_BUFFER_SIZE);
+    audio_buffer_read_index = 0;
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(microbit_audio_reset_obj, audio_reset);
+
+
+mp_obj_t audio_init(void) {
+    static bool initialised = false;
+    if (!initialised) {
+        audio_source_iter = NULL;
+        initialised = true;
+        //Allocate buffer
+        audio_buffer_ptr = (int8_t *)malloc(AUDIO_BUFFER_SIZE);
+        NVIC_DisableIRQ(TheTimer_IRQn);
+        TheTimer->POWER = 1;
+        NRF_TIMER_Type *timer = TheTimer;
+        timer_stop();
+        timer->TASKS_CLEAR = 1;
+        timer->MODE = TIMER_MODE_MODE_Timer;
+        timer->BITMODE = TIMER_BITMODE_BITMODE_16Bit << TIMER_BITMODE_BITMODE_Pos;
+        timer->PRESCALER = 0; //Full speed
+        timer->INTENCLR = TIMER_INTENCLR_COMPARE0_Msk | TIMER_INTENCLR_COMPARE1_Msk |
+                          TIMER_INTENCLR_COMPARE2_Msk | TIMER_INTENCLR_COMPARE3_Msk;
+        timer->SHORTS = 0;
+    }
+    return audio_reset();
+}
+MP_DEFINE_CONST_FUN_OBJ_0(audio___init___obj, audio_init);
+
+
+STATIC mp_obj_t start() {
+    audio_start();
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_0(microbit_audio_start_obj, start);
+
+STATIC mp_obj_t stop() {
+    audio_stop();
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_0(microbit_audio_stop_obj, stop);
+
+STATIC mp_obj_t play(mp_uint_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_source, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_wait,  MP_ARG_BOOL, {.u_bool = true} },
+        { MP_QSTR_pins,   MP_ARG_OBJ, {.u_obj = mp_const_none } },
+    };
+    // parse args
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+    mp_obj_t src = args[0].u_obj;
+    mp_obj_t pins = args[2].u_obj;
+    if (pins == mp_const_none) {
+        audio_set_pins((mp_obj_t)&microbit_p0_obj, (mp_obj_t)&microbit_p1_obj);
+    } else {
+        mp_obj_t len = mp_obj_len(pins);
+        if (mp_obj_get_int(len) != 2) {
+            nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "wrong number of pins"));
+        }
+        mp_obj_t pin0 = mp_obj_subscr(pins, MP_OBJ_NEW_SMALL_INT(0), MP_OBJ_SENTINEL);
+        mp_obj_t pin1 = mp_obj_subscr(pins, MP_OBJ_NEW_SMALL_INT(1), MP_OBJ_SENTINEL);
```

Comment:
There is an API function just for this pattern: `mp_obj_get_array_fixed_n(pins, 2, &items);`